seed/seed20.go

```diff
@@ -135,47 +175,155 @@ func (s *seed20) Copy(seedDir string, label string, tm timings.Measurer) (err er
-		if info.IsDir() {
-			return os.Mkdir(destPath, info.Mode())
-		}
-
-		return osutil.CopyFile(path, destPath, osutil.CopyFlagDefault)
-	})
+	entries, err := os.ReadDir(srcSystemDir)
 	if err != nil {
 		return err
 	}
 
-	destAssertedSnapDir := filepath.Join(destSeedDir, "snaps")
+	// copy all of the top-level files and directories from the seed. we'll copy
+	// the rest manually
+	for _, e := range entries {
+		if e.IsDir() {
+			if err := os.Mkdir(filepath.Join(destSystemDir, e.Name()), 0755); err != nil {
+				return err
+			}
+		} else {
+			if err := osutil.CopyFile(
+				filepath.Join(srcSystemDir, e.Name()),
+				filepath.Join(destSystemDir, e.Name()),
+				osutil.CopyFlagDefault,
+			); err != nil {
+				return err
+			}
+		}
+	}
 
-	if err := os.MkdirAll(destAssertedSnapDir, 0755); err != nil {
+	destAssertedSnapDir := filepath.Join(destSeedDir, "snaps")
+	if err := os.Mkdir(destAssertedSnapDir, 0755); err != nil {
 		return err
 	}
 
-	// copy the asserted snaps that the seed needs
+	destUnassertedSnapDir := filepath.Join(destSystemDir, "snaps")
+
+	// while copying the snaps, we also collect the assertions and the fields in
+	// the aux-info.json file that we need to write
+	var snapAssertions []asserts.Assertion
+	auxInfo := make(map[string]*internal.AuxInfo20)
+
+	// copy the snaps and components that the seed needs
 	for _, sn := range s.snaps {
-		// unasserted snaps are already copied above, skip them
-		if sn.ID() == "" {
+		// if we're not copying the snap, then we also don't need to copy the
+		// components for this snap
+		if !shouldCopySnap(sn, s.model, opts.OptionalContainers) {
 			continue
 		}
 
-		destSnapPath := filepath.Join(destAssertedSnapDir, filepath.Base(sn.Path))
+		// note that assertions should always be present if we have a snap ID
+		if a, ok := s.snapDeclsByID[sn.ID()]; ok {
+			snapAssertions = append(snapAssertions, a)
+		}
+		if a, ok := s.snapRevsByID[sn.ID()]; ok {
+			snapAssertions = append(snapAssertions, a)
+		}
+
+		if a, ok := s.auxInfos[sn.ID()]; ok {
+			auxInfo[sn.ID()] = a
+		}
+
+		var destSnapPath string
+		if sn.ID() == "" {
+			destSnapPath = filepath.Join(destUnassertedSnapDir, filepath.Base(sn.Path))
+		} else {
+			destSnapPath = filepath.Join(destAssertedSnapDir, filepath.Base(sn.Path))
+		}
 
 		if err := osutil.CopyFile(sn.Path, destSnapPath, osutil.CopyFlagOverwrite); err != nil {
 			return fmt.Errorf("cannot copy asserted snap: %w", err)
 		}
+
+		for _, comp := range sn.Components {
+			if !shouldCopyComponent(comp, sn.SnapName(), s.model, opts.OptionalContainers) {
+				continue
+			}
+
+			key := resourceKey{snapID: sn.ID(), name: comp.CompSideInfo.Component.ComponentName}
+			if a, ok := s.resPairByResKey[key]; ok {
+				snapAssertions = append(snapAssertions, a)
+			}
+			if a, ok := s.resRevByResKey[key]; ok {
+				snapAssertions = append(snapAssertions, a)
+			}
+
+			var destCompPath string
+			if !comp.CompSideInfo.Revision.Store() {
```

Comment:
For the snap, we check `sn.ID() == ""`, I wonder if we should check that here as well instead of the revision. As all components are expected to be (un)asserted if the snap is. So the relationship is clearer. Or maybe add a comment saying that both checks should produce the same result.